query-engine/black-box-tests/tests/protocols/mismatched.rs

```diff
@@ -0,0 +1,95 @@
+
+const JSON_QUERY: &'static str = r###"
+{
+    "action": "findMany",
+    "modelName": "Person",
+    "query": {
+        "arguments": {
+        },
+        "selection": {
+            "$scalars": true
+        }
+    }
+}
+"###;
+
+const GRAPHQL_QUERY: &'static str = r###"
+{
+  "operationName": null,
+  "variables": {},
+  "query": "{\n  findManyPerson {\n    id\n  }\n}\n"
+}
+"###;
+
+#[test_suite(schema(schema))]
+mod mismatched {
+    fn schema() -> String {
+        let schema = indoc! {
+            r#"model Person {
+              #id(id, Int, @id)
+             }"#
+        };
+
+        schema.to_owned()
+    }
+
+    #[connector_test]
+    async fn json_query_over_json_protocol_engine(r: Runner) -> TestResult<()> {
+        let mut qe_cmd = query_engine_cmd(r.prisma_dml(), "57582");
+        qe_cmd.env("PRISMA_ENGINE_PROTOCOL", "json");
+
+        with_child_process(&mut qe_cmd, async move {
+            let client = reqwest::Client::new();
+            let res = client
+                .post("http://0.0.0.0:57582/")
+                .body(JSON_QUERY)
+                .send()
+                .await
+                .unwrap();
+            insta::assert_snapshot!(res.text().await.unwrap(), @r###"{"data":{"findManyPerson":[]}}"###);
+        })
+        .await
+    }
+
+    #[connector_test]
+    async fn graphql_query_over_json_protocol_engine(r: Runner) -> TestResult<()> {
+        let mut qe_cmd = query_engine_cmd(r.prisma_dml(), "57582");
+        qe_cmd.env("PRISMA_ENGINE_PROTOCOL", "json");
+
+        with_child_process(&mut qe_cmd, async move {
+            let client = reqwest::Client::new();
+            let res = client
+                .post("http://0.0.0.0:57582/")
+                .body(GRAPHQL_QUERY)
+                .send()
+                .await
+                .unwrap();
+
+            assert_eq!(res.status(), reqwest::StatusCode::UNPROCESSABLE_ENTITY);
+            insta::assert_snapshot!(res.text().await.unwrap(), @r###"{"is_panic":false,"message":"Error parsing Json query. data did not match any variant of untagged enum JsonBody","backtrace":null}"###);
+        })
+        .await
+    }
+
+    #[connector_test]
+    async fn json_query_over_graphql_protocol_engine(r: Runner) -> TestResult<()> {
+        let mut qe_cmd = query_engine_cmd(r.prisma_dml(), "57582");
+        qe_cmd.env("PRISMA_ENGINE_PROTOCOL", "graphql");
+
+        with_child_process(&mut qe_cmd, async move {
+            let client = reqwest::Client::new();
+            let res = client
+                .post("http://0.0.0.0:57582/")
+                .body(JSON_QUERY)
+                .send()
+                .await
+                .unwrap();
+
+            assert_eq!(res.status(), reqwest::StatusCode::UNPROCESSABLE_ENTITY);
+            insta::assert_snapshot!(res.text().await.unwrap(), @r###"{"is_panic":false,"message":"Error parsing Graphql query. data did not match any variant of untagged enum GraphqlBody","backtrace":null}"###);
```

Comment:
Nice how the error message include the kind of query it is expecting and parsing for 👍